Fix #575: Low fi Home Page Greeting message

veena14cs opened the pull request:
## Explanation
This PR displays the name of the greeting message on home page.
- For testing on different timezone go to phone settings and change the timezone.
- In testcases I have added fake timings that is converted to UTC timezone which is also according to circle ci timezone.

### Mock
https://xd.adobe.com/spec/e2239cf4-9cde-4c08-5296-25316c1f0a14-9412/screen/ce2a96de-2c98-41f9-91e6-6ca0d8135fb2/HP-Home-Page-2

### Screenshot
![Screenshot_20191230-141815](https://user-images.githubusercontent.com/16301028/71610766-c1e25080-2bb9-11ea-9ceb-292f3c8232b1.png)



## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

veena14cs committed:
Display profile name

veena14cs commented:
@mschanteltc I am not finding specs of this mock. PTAL.

mschanteltc commented:
> @mschanteltc I am not finding specs of this mock. PTAL.

Correct me if I'm wrong, but this references "Good afternoon, Sean!" at the top of the page, but below the header.

veena14cs commented:
> > @mschanteltc I am not finding specs of this mock. PTAL.
> 
> Correct me if I'm wrong, but this references "Good afternoon, Sean!" at the top of the page, but below the header.

Yes we wanted to confirm this with you. Whether to display Welcome to Oppia or Good Afternoon message. I have added screenshot for your reference https://user-images.githubusercontent.com/16301028/71610766-c1e25080-2bb9-11ea-9ceb-292f3c8232b1.png

rt4914 reviewed (changes requested):
Please check this gist and update code accordingly.
The main reason behind introducing this function
```
private fun setProfileName() {
    if (::userAppHistoryViewModel.isInitialized && ::profileName.isInitialized) {
      userAppHistoryViewModel.profileName.set(" $profileName")
    }
  }
```

is that we cannot be sure whether the `subscribeToProfileLiveData` function will finish first or `subscribeToUserAppHistory`. So, we need to be sure that both the functions have finished before setting the profile name.

Also, do add test case for this functionality.

mschanteltc commented:
> > > @mschanteltc I am not finding specs of this mock. PTAL.
> > 
> > 
> > Correct me if I'm wrong, but this references "Good afternoon, Sean!" at the top of the page, but below the header.
> 
> Yes we wanted to confirm this with you. Whether to display Welcome to Oppia or Good Afternoon message. I have added screenshot for your reference https://user-images.githubusercontent.com/16301028/71610766-c1e25080-2bb9-11ea-9ceb-292f3c8232b1.png

Is it possible we can give a greeting based on the time? So "Good morning," "Good afternoon," and "Good evening," followed by the user's name.

nikitamarysolomanpvt reviewed (commented):
LGTM please assign me back once suggestions of mschanteltc are implemented.

rt4914 commented:
> > > > @mschanteltc I am not finding specs of this mock. PTAL.
> > > 
> > > 
> > > Correct me if I'm wrong, but this references "Good afternoon, Sean!" at the top of the page, but below the header.
> > 
> > 
> > Yes we wanted to confirm this with you. Whether to display Welcome to Oppia or Good Afternoon message. I have added screenshot for your reference https://user-images.githubusercontent.com/16301028/71610766-c1e25080-2bb9-11ea-9ceb-292f3c8232b1.png
> 
> Is it possible we can give a greeting based on the time? So "Good morning," "Good afternoon," and "Good evening," followed by the user's name.

@mschanteltc It is possible. Do you want us to do that?

veena14cs commented:
> > > > @mschanteltc I am not finding specs of this mock. PTAL.
> > > 
> > > 
> > > Correct me if I'm wrong, but this references "Good afternoon, Sean!" at the top of the page, but below the header.
> > 
> > 
> > Yes we wanted to confirm this with you. Whether to display Welcome to Oppia or Good Afternoon message. I have added screenshot for your reference https://user-images.githubusercontent.com/16301028/71610766-c1e25080-2bb9-11ea-9ceb-292f3c8232b1.png
> 
> Is it possible we can give a greeting based on the time? So "Good morning," "Good afternoon," and "Good evening," followed by the user's name.

Showing greeting based on time

![Screenshot_20200102-160237](https://user-images.githubusercontent.com/16301028/71662971-6eb6fc00-2d79-11ea-9d70-c7e7ba86af7f.png)

veena14cs committed:
changed greeting message

rt4914 commented:
@veena14cs We can hold this PR for now, there are some points which needs to be cleared before finishing this PR.

1. Greeting message -> Are we finalising time based greeting? 
2. What if the text name is big, then how should the UI change, because the text will take 2 lines. For example, `Good afternoon, Harshvardhan`

nikitamarysolomanpvt commented:
> @veena14cs We can hold this PR for now, there are some points which needs to be cleared before finishing this PR.
> 
> 1. Greeting message -> Are we finalising time based greeting?
> 2. What if the text name is big, then how should the UI change, because the text will take 2 lines. For example, `Good afternoon, Harshvardhan`

@veena14cs please finalise the requirement and assign me back for the final review

mschanteltc commented:
> > > > > @mschanteltc I am not finding specs of this mock. PTAL.
> > > > 
> > > > 
> > > > Correct me if I'm wrong, but this references "Good afternoon, Sean!" at the top of the page, but below the header.
> > > 
> > > 
> > > Yes we wanted to confirm this with you. Whether to display Welcome to Oppia or Good Afternoon message. I have added screenshot for your reference https://user-images.githubusercontent.com/16301028/71610766-c1e25080-2bb9-11ea-9ceb-292f3c8232b1.png
> > 
> > 
> > Is it possible we can give a greeting based on the time? So "Good morning," "Good afternoon," and "Good evening," followed by the user's name.
> 
> @mschanteltc It is possible. Do you want us to do that?

Yes, and looking at Veena's screenshot, it LGTM. Only correction to make is to remove the capitalization from the second word.

veena14cs committed:
Update strings.xml

veena14cs commented:
> > > > > > @mschanteltc I am not finding specs of this mock. PTAL.
> > > > > 
> > > > > 
> > > > > Correct me if I'm wrong, but this references "Good afternoon, Sean!" at the top of the page, but below the header.
> > > > 
> > > > 
> > > > Yes we wanted to confirm this with you. Whether to display Welcome to Oppia or Good Afternoon message. I have added screenshot for your reference https://user-images.githubusercontent.com/16301028/71610766-c1e25080-2bb9-11ea-9ceb-292f3c8232b1.png
> > > 
> > > 
> > > Is it possible we can give a greeting based on the time? So "Good morning," "Good afternoon," and "Good evening," followed by the user's name.
> > 
> > 
> > @mschanteltc It is possible. Do you want us to do that?
> 
> Yes, and looking at Veena's screenshot, it LGTM. Only correction to make is to remove the capitalization from the second word.

Removed capitalization from second word.
![Screenshot_20200103-120828](https://user-images.githubusercontent.com/16301028/71710357-db85d100-2e21-11ea-98a4-668b2353648e.png)

veena14cs commented on app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt:
```diff
@@ -79,12 +91,36 @@ class HomeFragmentPresenter @Inject constructor(
     return binding.root
   }
 
+
```
Remove extra spacing.

rt4914 commented on app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt:
```diff
@@ -148,4 +192,15 @@ class HomeFragmentPresenter @Inject constructor(
+      in 0..11 -> userAppHistoryViewModel.greeting = fragment.requireContext().getString(R.string.good_morning)
+      in 12..15 -> userAppHistoryViewModel.greeting = fragment.requireContext().getString(R.string.good_afternoon)
+      in 16..20 -> userAppHistoryViewModel.greeting = fragment.requireContext().getString(R.string.good_evening)
+      in 21..23 -> userAppHistoryViewModel.greeting = fragment.requireContext().getString(R.string.good_night)
```
Lets update these timings like this:

5-11 -> Morning
12 - 16 -> Afternoon
17 - 22 -> Evening
23 - 4 -> Night

Lets finish this for now, we can discuss this in meeting again and if there are changes it will act as a good started issue for a new developer.

veena14cs replied:
Okay, I have updated these timings.

rt4914 reviewed (approved):
One nit change:

in HomeFragmentPresenter -> `Options + Cmd + L` for fixing formatting

veena14cs commented:
> One nit change:
> 
> in HomeFragmentPresenter -> `Options + Cmd + L` for fixing formatting

Fixed formatting.

veena14cs committed:
Update HomeFragmentPresenter.kt

veena14cs committed:
working on testcase

rt4914 commented on app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt:
```diff
@@ -148,4 +193,15 @@ class HomeFragmentPresenter @Inject constructor(
+      in 5..11 -> userAppHistoryViewModel.greeting = fragment.requireContext().getString(R.string.good_morning)
+      in 12..16 -> userAppHistoryViewModel.greeting = fragment.requireContext().getString(R.string.good_afternoon)
+      in 17..22 -> userAppHistoryViewModel.greeting = fragment.requireContext().getString(R.string.good_evening)
+      in 23 downTo 4 -> userAppHistoryViewModel.greeting = fragment.requireContext().getString(R.string.good_night)
```
Replace Goodnight to Goodevening

veena14cs replied:
Done.

rt4914 commented:
@veena14cs just noticed few more things:

1. The PR title is not correct:
`Fix #575 : ....`

2. Checklist in PR description is not filled.

veena14cs committed:
fetching correct id

veena14cs committed:
working on testcase

veena14cs committed:
Update HomeFragmentPresenter.kt

veena14cs committed:
Update HomeActivityTest.kt

veena14cs committed:
updated testcase

veena14cs committed:
Update HomeActivityTest.kt

veena14cs committed:
updated test cases

veena14cs committed:
Update HomeActivity.kt

veena14cs committed:
nit change

veena14cs committed:
Merge branch 'develop' into hifi-welcome-profile-name

veena14cs committed:
Update ProfileManagementController.kt

veena14cs commented:
> @veena14cs just noticed few more things:
> 
> 1. The PR title is not correct:
>    `Fix #575 : ....`
> 2. Checklist in PR description is not filled.

updated.

rt4914 commented on app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt:
```diff
@@ -57,6 +68,10 @@ class HomeFragmentPresenter @Inject constructor(
 
+    id = activity.intent.getIntExtra(KEY_HOME_PROFILE_ID, 0)
+    profileId = ProfileId.newBuilder().setInternalId(id).build()
+//    profileId = profileManagementController.getCurrentProfileId()
```
Remove this line

veena14cs replied:
Removed.

rt4914 commented on app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt:
```diff
@@ -57,6 +68,10 @@ class HomeFragmentPresenter @Inject constructor(
     itemList.add(allTopicsViewModel)
     topicListAdapter = TopicListAdapter(activity, itemList, promotedStoryList)
 
+    id = activity.intent.getIntExtra(KEY_HOME_PROFILE_ID, 0)
```
Why are we still passing id via intent, we should be using `profileManagementController.getCurrentProfileId()`

veena14cs replied:
With intent it returns correct id, where as if we use profileManagementController.getCurrentProfileId() it is giving -1 in testcases. To fix that we are using id through intent.

rt4914 commented on app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt:
```diff
@@ -79,12 +94,35 @@ class HomeFragmentPresenter @Inject constructor(
+
+  private fun processGetProfileResult(profileResult: AsyncResult<Profile>): Profile {
+    if (profileResult.isFailure()) {
+      logger.e("OptionsFragment", "Failed to retrieve profile", profileResult.getErrorOrNull()!!)
```
Incorrect tag.

veena14cs replied:
Updated.

rt4914 commented on app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt:
```diff
@@ -125,6 +164,14 @@ class HomeFragmentPresenter @Inject constructor(
+    if (::userAppHistoryViewModel.isInitialized && ::profileName.isInitialized) {
+      displayGreeting()
+      userAppHistoryViewModel.profileName = "$profileName!"
+      logger.d("profile /id", "==" + profileId.internalId)
```
Remove this line

veena14cs replied:
Removed.

rt4914 commented on app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt:
```diff
@@ -148,4 +195,15 @@ class HomeFragmentPresenter @Inject constructor(
+      in 5..11 -> userAppHistoryViewModel.greeting = fragment.requireContext().getString(R.string.good_morning)
+      in 12..16 -> userAppHistoryViewModel.greeting = fragment.requireContext().getString(R.string.good_afternoon)
+      in 17..22 -> userAppHistoryViewModel.greeting = fragment.requireContext().getString(R.string.good_evening)
+      in 23 downTo 4 -> userAppHistoryViewModel.greeting = fragment.requireContext().getString(R.string.good_evening)
```
Last two conditions should be merged such that `in  17 downTo 4`

veena14cs replied:
updated.

rt4914 commented on app/src/sharedTest/java/org/oppia/app/home/HomeActivityTest.kt:
```diff
@@ -85,46 +95,67 @@ class HomeActivityTest {
+          ).check(matches(withText("Good evening,")))
+        }
+
+        else -> onView(
```
Update this test cases as per implementation change update

veena14cs replied:
Updated.

rt4914 commented on app/src/sharedTest/java/org/oppia/app/home/HomeActivityTest.kt:
```diff
@@ -454,6 +474,8 @@ class HomeActivityTest {
     }
 
     fun getUserAppHistoryController(): UserAppHistoryController
```
Remove the usage of userAppHistory and its controller

veena14cs replied:
removed.

rt4914 commented on app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt:
```diff
@@ -18,20 +19,26 @@ import org.oppia.app.home.topiclist.TopicListAdapter
   private val activity: AppCompatActivity,
   private val fragment: Fragment,
+  private val profileManagementController: ProfileManagementController,
   private val userAppHistoryController: UserAppHistoryController,
```
Remove the usage of userAppHistory and its controller

veena14cs replied:
Its been used couple of times in the same class, not sure whether it is required in future. I will raise an issue related to this to remove unwanted userAppHistoryController #600 .

rt4914 replied:
Sounds good. Thanks.

rt4914 reviewed (changes requested):
PTAL.

nikitamarysolomanpvt commented on app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt:
```diff
@@ -57,6 +68,10 @@ class HomeFragmentPresenter @Inject constructor(
     itemList.add(allTopicsViewModel)
     topicListAdapter = TopicListAdapter(activity, itemList, promotedStoryList)
 
+    id = activity.intent.getIntExtra(KEY_HOME_PROFILE_ID, 0)
```
rename variable

veena14cs replied:
Done.

nikitamarysolomanpvt commented on app/src/main/res/values/strings.xml:
```diff
@@ -196,4 +196,8 @@
   <string name="profile_reset_pin_admin_confirm_label">Confirm 5-Digit PIN</string>
   <string name="profile_reset_pin_error_user_pin_length">Your PIN should be 3 digits long.</string>
   <string name="profile_reset_pin_error_admin_pin_length">Your PIN should be 5 digits long.</string>
+  <!-- Welcome -->
```
Please update this for user understanding or remove.

veena14cs replied:
Done.

nikitamarysolomanpvt reviewed (commented):
Please assign me back once all comments are resolved.

veena14cs committed:
fixed issues

veena14cs committed:
Update HomeActivityTest.kt

rt4914 commented on app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt:
```diff
@@ -44,6 +51,10 @@ class HomeFragmentPresenter @Inject constructor(
   private lateinit var allTopicsViewModel: AllTopicsViewModel
   private lateinit var topicListAdapter: TopicListAdapter
   private lateinit var binding: HomeFragmentBinding
+  private var profileIdFromIntent: Int = -1
```
Lets make this convention clear to avoid any confusion.

If the id is `int` we can name that as `internalProfileId`
and if it is of type `ProfileId` we can directly name that as `profileId`.

veena14cs replied:
Updated to internalProfileId

rt4914 commented on app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt:
```diff
@@ -57,6 +68,9 @@ class HomeFragmentPresenter @Inject constructor(
     itemList.add(allTopicsViewModel)
     topicListAdapter = TopicListAdapter(activity, itemList, promotedStoryList)
 
+    profileIdFromIntent = activity.intent.getIntExtra(KEY_HOME_PROFILE_ID, 0)
```
Default value should be `-1`. As `0` is related to admin profile.

veena14cs replied:
Updated to -1.

rt4914 commented on app/src/main/res/values/strings.xml:
```diff
@@ -196,4 +196,8 @@
   <string name="profile_reset_pin_admin_confirm_label">Confirm 5-Digit PIN</string>
   <string name="profile_reset_pin_error_user_pin_length">Your PIN should be 3 digits long.</string>
   <string name="profile_reset_pin_error_admin_pin_length">Your PIN should be 5 digits long.</string>
+  <!-- HomeFragmentPresenter. -->
```
Rename this as `<!-- HomeFragment -->` also, we will be removing that unnecessary `.` from all comments in this file.

veena14cs replied:
Done.

rt4914 commented on app/src/sharedTest/java/org/oppia/app/home/HomeActivityTest.kt:
```diff
@@ -53,30 +53,39 @@ import org.oppia.app.recyclerview.RecyclerViewMatcher.Companion.atPosition
 
 /** Tests for [HomeActivity]. */
 @RunWith(AndroidJUnit4::class)
 class HomeActivityTest {
```
I think the test cases needs to mock calendar so that all the test cases work.

nikitamarysolomanpvt replied:
+1

veena14cs replied:
Created a wrapper class and added fake time to test all 3 cases.

rt4914 commented on app/src/sharedTest/java/org/oppia/app/home/HomeActivityTest.kt:
```diff
@@ -85,62 +94,90 @@ class HomeActivityTest {
+              R.id.welcome_text_view
+            )
+          ).check(matches(withText("Good afternoon,")))
+        }
```
Add else condition in all test cases in which test case will fail everytime.

veena14cs replied:
I have changed this test. PTAL.

rt4914 reviewed (changes requested):
Suggested some changes

nikitamarysolomanpvt reviewed (commented):
LGTM Please add dedicated test-case Activity and assign me back.

veena14cs committed:
made util class for date time

veena14cs committed:
Fixed testcases.

veena14cs committed:
Update strings.xml

rt4914 commented on app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt:
```diff
@@ -148,4 +193,14 @@ class HomeFragmentPresenter @Inject constructor(
+//      in 12..16 -> userAppHistoryViewModel.greeting = fragment.requireContext().getString(R.string.good_afternoon)
+//      in 17 downTo 4 -> userAppHistoryViewModel.greeting = fragment.requireContext().getString(R.string.good_evening)
+//    }
+//  }
```
Remove this commented code.

veena14cs replied:
Removed.

rt4914 commented on app/src/sharedTest/java/org/oppia/app/home/HomeActivityTest.kt:
```diff
@@ -53,30 +55,44 @@ import org.oppia.app.recyclerview.RecyclerViewMatcher.Companion.atPosition
+  @Inject
+  lateinit var context: Context
+  @Inject
+  lateinit var dateTimeUtil: DateTimeUtil
```
These should be single lined:

```
  @Inject lateinit var context: Context
  @Inject lateinit var dateTimeUtil: DateTimeUtil
  @Inject lateinit var profileTestHelper: ProfileTestHelper
```

veena14cs replied:
Done.

rt4914 commented on app/src/sharedTest/java/org/oppia/app/home/HomeActivityTest.kt:
```diff
@@ -85,67 +101,65 @@ class HomeActivityTest {
-  fun testHomeActivity_secondOpen_hasWelcomeBackString() {
-    simulateAppAlreadyOpened()
+  fun testHomeActivity_recyclerViewIndex0_getGreetingMessageBasedOnTime_goodMorningMessageSuccessful() {
+    launch<HomeActivity>(createHomeActivityIntent(0)).use {
```
This test has nothing to do with recyclerview or the home activity. These test cases should go in utility module and in here you should check whether the text is getting displayed correctly or not.

veena14cs replied:
I have updated testcases.

rt4914 reviewed (changes requested):
Suggested changes and updates in test cases

veena14cs committed:
Create DateTimeUtilTest.kt

nikitamarysolomanpvt reviewed (approved):
LGTM Please confirm with @rajat the naming conversion of string res.

veena14cs committed:
working on mocking calender

rt4914 commented:
> LGTM Please confirm with @rajat the naming conversion of string res.

@nikitamarysolomanpvt I think the naming convention is fine in this PR for string resources.

veena14cs committed:
created oppia clock

veena14cs committed:
working on testcase

veena14cs committed:
updated testcase

veena14cs committed:
Update HomeActivityTest.kt

veena14cs committed:
Update HomeFragmentPresenter.kt

veena14cs commented:
@BenHenning @rt4914 @nikitamarysolomanpvt To set fake times for testing different greeting messages I have made use of `OppiaClock` class in utility module. To have same instance of this class in test and production code, I have injected `OppiaClock` in `HomeActivity` and its instance is getting passed through all the classes to reach `HomeFragmentPresenter` that uses this instance. 

With this approach all the testcases are passing, but its not a cleaner way to do it. And I need suggestion on how to use a cleaner way to get one instance of `OppiaClock` in test and Production code or to tweak fake time for testing.

veena14cs committed:
nit change

nikitamarysolomanpvt commented:
> > LGTM Please confirm with @rajat the naming conversion of string res.
> 
> @nikitamarysolomanpvt I think the naming convention is fine in this PR for string resources.

Ya with the updated comment in string res it looks good to me also.

rt4914 reviewed (approved):
LGTM, but will need to defer to @BenHenning for a cleaner approach for test cases using OppiaClock.

veena14cs committed:
fixed testcase using homeInjectionActivity

veena14cs commented:
> LGTM, but will need to defer to @BenHenning for a cleaner approach for test cases using OppiaClock.

Changed this approach. Introduced HomeInjectionTestActivity to inject oppiaClock for testcases as suggested by @BenHenning . PTAL once again.

veena14cs committed:
fixed nit changes

rt4914 commented on app/src/main/java/org/oppia/app/home/HomeActivity.kt:
```diff
@@ -1,18 +1,30 @@
 import org.oppia.app.activity.InjectableAppCompatActivity
 import org.oppia.app.profile.ProfileActivity
 import org.oppia.app.topic.TopicActivity
+import org.oppia.util.system.OppiaClock
```
Remove unused import

veena14cs replied:
Removed.

rt4914 commented on app/src/main/java/org/oppia/app/home/HomeActivityPresenter.kt:
```diff
@@ -9,6 +9,8 @@ import org.oppia.app.activity.ActivityScope
 import org.oppia.app.drawer.NavigationDrawerFragment
 import javax.inject.Inject
 
+const val TAG_HOME_FRAGMENT = "TAG_HOME_FRAGMENT"
```
It should like this:
`const val TAG_HOME_FRAGMENT = "HOME_FRAGMENT"`

veena14cs replied:
updated.

rt4914 commented on app/src/main/java/org/oppia/app/home/HomeFragmentPresenter.kt:
```diff
@@ -125,6 +163,13 @@ class HomeFragmentPresenter @Inject constructor(
+  private fun setProfileName() {
+    if (::userAppHistoryViewModel.isInitialized && ::profileName.isInitialized) {
+      userAppHistoryViewModel.profileName = "$profileName!"
+      userAppHistoryViewModel.greeting = DateTimeUtil(fragment.requireContext(),oppiaClock).getGreetingMessage()
```
Spacing issue: `userAppHistoryViewModel.greeting = DateTimeUtil(fragment.requireContext(), oppiaClock).getGreetingMessage()`

veena14cs replied:
Fixed.

rt4914 commented on app/src/sharedTest/java/org/oppia/app/home/HomeActivityTest.kt:
```diff
@@ -51,32 +52,45 @@ import org.oppia.app.home.continueplaying.ContinuePlayingActivity
 class HomeActivityTest {
 
+  @Inject lateinit var profileTestHelper: ProfileTestHelper
+  @Inject lateinit var profileManagementController: ProfileManagementController
```
Remove this unused line

veena14cs replied:
done.

rt4914 commented on app/src/sharedTest/java/org/oppia/app/home/HomeActivityTest.kt:
```diff
@@ -85,67 +99,97 @@ class HomeActivityTest {
-    simulateAppAlreadyOpened()
+  fun testHomeActivity_recyclerViewIndex0_displayGreetingMessageBasedOnTime_goodMorningMessageDisplayedSuccessful() {
+    getApplicationDependencies()
+    oppiaClock.setCurrentTimeMs(1579666500000)
```
Let makes this code cleaner. At the top of this file, create this as constant:

`private const val MORNING_TIMESTAMP = 1579666500000 `

and use `MORNING_TIMESTAMP` at this place. Similarly in other cases too.

veena14cs replied:
DONE.

rt4914 commented on utility/src/main/java/org/oppia/util/datetime/DateTimeUtil.kt:
```diff
@@ -0,0 +1,24 @@
+
+/** Utility to manage date and time. */
+@Singleton
+class DateTimeUtil @Inject constructor(private val context: Context,
```
Formatting fix:
```
class DateTimeUtil @Inject constructor(
  private val context: Context,
  private val oppiaClock: OppiaClock
) {
```

veena14cs replied:
Fixed

rt4914 commented on utility/src/main/java/org/oppia/util/system/OppiaClock.kt:
```diff
@@ -0,0 +1,32 @@
+
+  fun getCurrentCalendar(): Calendar {
+    val calendar = Calendar.getInstance()
+    calendar.setTimeInMillis(getCurrentTimeMs())
```
Update: `calendar.timeInMillis = getCurrentTimeMs()`

veena14cs replied:
DONE.

rt4914 commented on utility/src/test/java/org/oppia/util/datetime/DateTimeUtilTest.kt:
```diff
@@ -0,0 +1,154 @@
+
+  @Test
+  fun testGreetingMessageBasedOnTime_goodMorningMessageSucceed() {
+    oppiaClock.setCurrentTimeMs(1579666500000)
```
Create constants for this file too, as mentioned in `HomeActivityTest`

veena14cs replied:
Done.

rt4914 reviewed (changes requested):
PTAL

veena14cs committed:
fixed issues

rt4914 reviewed (approved):
LGTM, Thanks.

rt4914 commented:
@veena14cs Check CircleCI, these test cases are actually passing in android studio.

rt4914 commented:
<img width="915" alt="Screenshot 2020-01-23 at 2 11 12 PM" src="https://user-images.githubusercontent.com/9396084/72968935-8585c700-3dea-11ea-9475-9f5776259aaa.png">

veena14cs committed:
Update DateTimeUtilTest.kt

veena14cs committed:
Update DateTimeUtilTest.kt

nikitamarysolomanpvt commented on utility/src/main/java/org/oppia/util/system/OppiaClock.kt:
```diff
@@ -0,0 +1,32 @@
+  fun getCurrentCalendar(): Calendar {
+    val calendar = Calendar.getInstance()
+    calendar.timeInMillis = getCurrentTimeMs()
+    return calendar
```
You may directly return `Calendar.getInstance().timeInMillis=getCurrentTimeMs()`

veena14cs replied:
we cannot do assignment in return..

nikitamarysolomanpvt commented on utility/src/main/java/org/oppia/util/system/OppiaClock.kt:
```diff
@@ -0,0 +1,32 @@
+    testTimeMs = currentTimeMs
+  }
+
+  fun getCurrentCalendar(): Calendar {
```
Please add Kdoc.

veena14cs replied:
added.

veena14cs committed:
Update DateTimeUtilTest.kt

nikitamarysolomanpvt commented on utility/src/test/java/org/oppia/util/datetime/DateTimeUtilTest.kt:
```diff
@@ -0,0 +1,158 @@
+  }
+
+  @Test
+  fun testGreetingMessageBasedOnTime_goodMorningMessageSucceed() {
```
"Succeed" update to "SucceededResult" or "Succeeded"

veena14cs replied:
used Succeeded.

nikitamarysolomanpvt reviewed (approved):
LGTM Some nit suggestions PTAL

veena14cs committed:
nit change

veena14cs committed:
Update OppiaClock.kt

veena14cs committed:
update to circle ci time zone

veena14cs committed:
changed time condition

veena14cs committed:
added log

veena14cs committed:
update gradle

veena14cs committed:
Update DateTimeUtilTest.kt

veena14cs committed:
removed logging

veena14cs committed:
Merge branch 'develop' into hifi-welcome-profile-name

veena14cs committed:
Merge branch 'develop' into hifi-welcome-profile-name

veena14cs committed:
nit change

veena14cs committed:
Update DateTimeUtilTest.kt

veena14cs committed:
Merge branch 'develop' into hifi-welcome-profile-name

veena14cs committed:
Update WelcomeViewModel.kt

veena14cs committed:
updated time stamp.

veena14cs committed:
Update DateTimeUtilTest.kt

veena14cs committed:
Update DateTimeUtilTest.kt

veena14cs committed:
Update DateTimeUtilTest.kt

veena14cs committed:
add test method

veena14cs committed:
Merge branch 'develop' into hifi-welcome-profile-name

veena14cs committed:
Update DateTimeUtilTest.kt

veena14cs committed:
fix formatting.

veena14cs committed:
Update DateTimeUtilTest.kt

veena14cs committed:
Fixed test issues

veena14cs committed:
fixed nit issues

veena14cs commented:
Please test it by changing to different timezone.

rt4914 commented on app/src/main/java/org/oppia/app/home/WelcomeViewModel.kt:
```diff
@@ -1,9 +1,9 @@
 /** [ViewModel] for welcome text in home screen. */
 class WelcomeViewModel : HomeItemViewModel() {
   var profileName : String = ""
+  var greeting : String = ""
```
Prefer converting these strings  to `val` by making them observable field or pass them directly in the constructor.

nikitamarysolomanpvt replied:
+1

veena14cs replied:
Done.

rt4914 reviewed (approved):
`UserAppHistoryViewModel` file can be removed.

Nit changes suggested.

Also
`testHomeActivity_recyclerViewIndex0_configurationChange_displaysWelcomeMessageCorrectly()`
is failing on Nexus S - HDPI device. And on pixel it is passing. So it would be nice if you can fix this, if it does not get fixed its fine.

rt4914 commented on utility/src/main/java/org/oppia/util/datetime/DateTimeUtil.kt:
```diff
@@ -0,0 +1,27 @@
+
+  fun getGreetingMessage(): String {
+    val calender = oppiaClock.getCurrentCalendar()
+    System.out.println("date util=" + calender.timeInMillis)
```
Remove this

veena14cs replied:
Done.

rt4914 commented on app/src/sharedTest/java/org/oppia/app/home/HomeActivityTest.kt:
```diff
@@ -81,6 +83,10 @@ import java.util.concurrent.TimeoutException
+private const val AFTERNOON_TIMESTAMP = 1556029320000
+
 /** Tests for [HomeActivity]. */
 @RunWith(AndroidJUnit4::class)
```
Press `Control + Alt + L` - reformatting shortcut to fix the formatting.

veena14cs replied:
Done.

rt4914 commented on utility/src/main/java/org/oppia/util/system/OppiaClock.kt:
```diff
@@ -0,0 +1,61 @@
+    try {
+      date = format.parse(dtStart)
+      currentTimeMsNew = getLocalToUTCDate(date)
+      System.out.println("UTC = " + getLocalToUTCDate(date))
```
Remove this

veena14cs replied:
Done.

rt4914 commented on utility/src/main/java/org/oppia/util/system/OppiaClock.kt:
```diff
@@ -0,0 +1,61 @@
+    val time = calendar.time
+    val outputFmt = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
+    outputFmt.timeZone = TimeZone.getTimeZone("UTC")
+    System.out.println("time=" + outputFmt.format(time))
```
Remove this

veena14cs replied:
Done.

rt4914 reviewed (changes requested):
Missed on utility module in previous review. Can you mention the steps to test the `DateTimeUtilTest` file because I ran it on Android JUnit and it failed.

nikitamarysolomanpvt commented on app/src/main/AndroidManifest.xml:
```diff
@@ -117,6 +117,8 @@
     <activity android:name=".testing.ExplorationInjectionActivity" />
     <activity android:name=".testing.ExplorationTestActivity" />
     <activity android:name=".testing.HtmlParserTestActivity" />
+    <activity android:name=".testing.HomeInjectionActivity" />
```
Please arrange in alphabetical order.

veena14cs replied:
Done.

nikitamarysolomanpvt commented on app/src/main/java/org/oppia/app/testing/HomeTestActivityPresenter.kt:
```diff
@@ -0,0 +1,16 @@
+import org.oppia.app.activity.ActivityScope
+import javax.inject.Inject
+
+/** The presenter for [HomeTestActivityPresenter]. */
```
Please correct it to` /** The presenter for [HomeTestActivity]. */`

veena14cs replied:
Done.

nikitamarysolomanpvt commented on app/src/sharedTest/java/org/oppia/app/home/HomeActivityTest.kt:
```diff
@@ -81,6 +83,10 @@ import java.util.concurrent.TimeoutException
 
+private const val EVENING_TIMESTAMP = 1556061720000
+private const val  MORNING_TIMESTAMP = 1556094120000
+private const val AFTERNOON_TIMESTAMP = 1556029320000
```
Please include doc about  time stamp  (the time in h:mm:ss)

veena14cs replied:
Done.

nikitamarysolomanpvt commented on utility/src/test/java/org/oppia/util/datetime/DateTimeUtilTest.kt:
```diff
@@ -0,0 +1,157 @@
+
+  // https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/
+  @ObsoleteCoroutinesApi
+  private val testThread = newSingleThreadContext("TestMain")
```
change code to ` @ObsoleteCoroutinesApi private val testThread = newSingleThreadContext("TestMain")`
 also Nit: ``Reference document: https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/``

veena14cs replied:
Updated.

nikitamarysolomanpvt commented on utility/src/test/java/org/oppia/util/datetime/DateTimeUtilTest.kt:
```diff
@@ -0,0 +1,157 @@
+  }
+
+  @Qualifier
+  annotation class TestDispatcher
```
Please change code to `` @Qualifier annotation class TestDispatcher``

veena14cs replied:
Done,

nikitamarysolomanpvt reviewed (changes requested):
Please address the changes suggested.

veena14cs commented:
> Missed on utility module in previous review. Can you mention the steps to test the `DateTimeUtilTest` file because I ran it on Android JUnit and it failed.

Can you post screenshot of failure. It is same as how we usually run roboelectric tests.

veena14cs committed:
Fixed issues

rt4914 commented:
@veena14cs PTAL. I think there is some mistake in the steps that I am following to run these test cases. 

<img width="1680" alt="Screenshot 2020-04-07 at 3 55 14 PM" src="https://user-images.githubusercontent.com/9396084/78658747-3418ae80-78e8-11ea-9001-20ca35c90f94.png">

veena14cs commented:
> @veena14cs PTAL. I think there is some mistake in the steps that I am following to run these test cases.
> 
> <img alt="Screenshot 2020-04-07 at 3 55 14 PM" width="1680" src="https://user-images.githubusercontent.com/9396084/78658747-3418ae80-78e8-11ea-9001-20ca35c90f94.png">

Did you try rebuilding or by doing invalidate cache and restart.

rt4914 reviewed (approved):
LGTM, the test cases passed. Just fix all the nit changes suggested earlier and its good to merge.

veena14cs commented:
> LGTM, the test cases passed. Just fix all the nit changes suggested earlier and its good to merge.

Fixed all nit changes.

nikitamarysolomanpvt reviewed (approved):
LGTM.
